File: src/wasm/tests/test_edit.c
// test_edit.c — Tests for engine_edit.c functions
#include "test_framework.h"
#include "../engine_core.h"
#include "../engine_edit.h"
#include <string.h>

extern EngineState g_state;

// ============ Helper ============

static void init_state(void) {
    memset(&g_state, 0, sizeof(g_state));
    engine_core_init(&g_state);
}

// ============ engine_toggle_event ============

static void test_toggle_add_event(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    CU_ASSERT(idx >= 0);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].event_count, 1);
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].events[idx].row, 10);
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].events[idx].position, 0);
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].events[idx].length, 120);
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].events[idx].enabled, 1);
}

static void test_toggle_remove_event(void) {
    init_state();
    engine_toggle_event(&g_state, 10, 0, 120);
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    CU_ASSERT_EQUAL(idx, -1);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].event_count, 0);
}

static void test_toggle_multiple_events(void) {
    init_state();
    engine_toggle_event(&g_state, 10, 0, 120);
    engine_toggle_event(&g_state, 12, 120, 120);
    engine_toggle_event(&g_state, 14, 240, 120);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].event_count, 3);
}

static void test_toggle_event_default_values(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 5, 0, 120);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    NoteEvent_C* ev = &g_state.patterns[ch][pat].events[idx];
    CU_ASSERT_EQUAL(ev->repeat_amount, 1);
    CU_ASSERT_EQUAL(ev->chord_amount, 1);
    CU_ASSERT_EQUAL(ev->chord_space, 2);
    CU_ASSERT_EQUAL(ev->arp_style, ARP_CHORD);
    CU_ASSERT_EQUAL(ev->sub_modes[SM_VELOCITY].values[0], 100);
    CU_ASSERT_EQUAL(ev->sub_modes[SM_VELOCITY].length, 1);
    CU_ASSERT_EQUAL(ev->sub_modes[SM_HIT].values[0], 100);
}

// ============ engine_remove_event ============

static void test_remove_event_fixes_selection(void) {
    init_state();
    engine_toggle_event(&g_state, 10, 0, 120);
    engine_toggle_event(&g_state, 12, 120, 120);
    g_state.selected_event_idx = 1;
    engine_remove_event(&g_state, 0);
    CU_ASSERT_EQUAL(g_state.selected_event_idx, 0);
}

static void test_remove_event_clears_selection(void) {
    init_state();
    engine_toggle_event(&g_state, 10, 0, 120);
    g_state.selected_event_idx = 0;
    engine_remove_event(&g_state, 0);
    CU_ASSERT_EQUAL(g_state.selected_event_idx, -1);
}

// ============ engine_move_event ============

static void test_move_event_basic(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    engine_move_event(&g_state, (uint16_t)idx, 20, 240);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].events[idx].row, 20);
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].events[idx].position, 240);
}

static void test_move_event_out_of_range_ignored(void) {
    init_state();
    engine_move_event(&g_state, 99, 20, 240); // should not crash
}

// ============ engine_set_event_length ============

static void test_set_length_basic(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    engine_set_event_length(&g_state, (uint16_t)idx, 240);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].events[idx].length, 240);
}

static void test_set_length_min_clamp(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    engine_set_event_length(&g_state, (uint16_t)idx, 0);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].events[idx].length, 1);
}

// ============ Sub-mode operations ============

static void test_sub_mode_value_basic(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    engine_set_sub_mode_value(&g_state, (uint16_t)idx, SM_VELOCITY, 0, 80);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].events[idx].sub_modes[SM_VELOCITY].values[0], 80);
}

static void test_sub_mode_value_materializes_array(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    engine_set_sub_mode_value(&g_state, (uint16_t)idx, SM_VELOCITY, 2, 50);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    SubModeArray* arr = &g_state.patterns[ch][pat].events[idx].sub_modes[SM_VELOCITY];
    CU_ASSERT_EQUAL(arr->length, 3);
    CU_ASSERT_EQUAL(arr->values[2], 50);
    CU_ASSERT_EQUAL(arr->values[1], 100); // looped from index 0
}

static void test_sub_mode_length_expand(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    engine_set_sub_mode_length(&g_state, (uint16_t)idx, SM_VELOCITY, 4);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].events[idx].sub_modes[SM_VELOCITY].length, 4);
}

static void test_sub_mode_length_shrink(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    engine_set_sub_mode_length(&g_state, (uint16_t)idx, SM_VELOCITY, 4);
    engine_set_sub_mode_length(&g_state, (uint16_t)idx, SM_VELOCITY, 2);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].events[idx].sub_modes[SM_VELOCITY].length, 2);
}

static void test_sub_mode_length_clamp_min(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    engine_set_sub_mode_length(&g_state, (uint16_t)idx, SM_VELOCITY, 0);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].events[idx].sub_modes[SM_VELOCITY].length, 1);
}

static void test_toggle_loop_mode_cycles(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    SubModeArray* arr = &g_state.patterns[ch][pat].events[idx].sub_modes[SM_VELOCITY];

    CU_ASSERT_EQUAL(arr->loop_mode, LOOP_RESET);
    engine_toggle_sub_mode_loop_mode(&g_state, (uint16_t)idx, SM_VELOCITY);
    CU_ASSERT_EQUAL(arr->loop_mode, LOOP_CONTINUE);
    engine_toggle_sub_mode_loop_mode(&g_state, (uint16_t)idx, SM_VELOCITY);
    CU_ASSERT_EQUAL(arr->loop_mode, LOOP_FILL);
    engine_toggle_sub_mode_loop_mode(&g_state, (uint16_t)idx, SM_VELOCITY);
    CU_ASSERT_EQUAL(arr->loop_mode, LOOP_RESET);
}

// ============ Chord operations ============

static void test_chord_stack_increase(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    NoteEvent_C* ev = &g_state.patterns[ch][pat].events[idx];

    CU_ASSERT_EQUAL(ev->chord_amount, 1);
    engine_adjust_chord_stack(&g_state, (uint16_t)idx, 1);
    CU_ASSERT_EQUAL(ev->chord_amount, 2);
    engine_adjust_chord_stack(&g_state, (uint16_t)idx, 1);
    CU_ASSERT_EQUAL(ev->chord_amount, 3);
}

static void test_chord_stack_decrease_clamp(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    NoteEvent_C* ev = &g_state.patterns[ch][pat].events[idx];

    engine_adjust_chord_stack(&g_state, (uint16_t)idx, -1);
    CU_ASSERT_EQUAL(ev->chord_amount, 1);
}

static void test_chord_stack_max_clamp(void) {
    init_state();
    int16_t idx = engine_toggle_event(&g_state, 10, 0, 120);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    NoteEvent_C* ev = &g_state.patterns[ch][pat].events[idx];

    for (int i = 0; i < 20; i++)
        engine_adjust_chord_stack(&g_state, (uint16_t)idx, 1);
    CU_ASSERT(ev->chord_amount <= MAX_CHORD_SIZE);
}

// ============ Pattern operations ============

static void test_copy_pattern(void) {
    init_state();
    engine_toggle_event(&g_state, 10, 0, 120);
    engine_toggle_event(&g_state, 12, 120, 120);
    engine_copy_pattern(&g_state, 1);
    uint8_t ch = g_state.current_channel;
    CU_ASSERT_EQUAL(g_state.patterns[ch][1].event_count, 2);
    CU_ASSERT_EQUAL(g_state.patterns[ch][1].events[0].row, 10);
    CU_ASSERT_EQUAL(g_state.patterns[ch][1].events[1].row, 12);
}

static void test_clear_pattern(void) {
    init_state();
    engine_toggle_event(&g_state, 10, 0, 120);
    engine_toggle_event(&g_state, 12, 120, 120);
    uint8_t ch = g_state.current_channel;
    uint8_t pat = g_state.current_patterns[ch];
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].event_count, 2);
    engine_clear_pattern(&g_state);
    CU_ASSERT_EQUAL(g_state.patterns[ch][pat].event_count, 0);
}

// ============ Suite registration ============

void register_edit_tests(void) {
    CU_pSuite suite = CU_add_suite("engine_edit", NULL, NULL);

    CU_add_test(suite, "toggle_add_event", test_toggle_add_event);
    CU_add_test(suite, "toggle_remove_event", test_toggle_remove_event);
    CU_add_test(suite, "toggle_multiple_events", test_toggle_multiple_events);
    CU_add_test(suite, "toggle_event_default_values", test_toggle_event_default_values);

    CU_add_test(suite, "remove_event_fixes_selection", test_remove_event_fixes_selection);
    CU_add_test(suite, "remove_event_clears_selection", test_remove_event_clears_selection);

    CU_add_test(suite, "move_event_basic", test_move_event_basic);
    CU_add_test(suite, "move_event_out_of_range_ignored", test_move_event_out_of_range_ignored);

    CU_add_test(suite, "set_length_basic", test_set_length_basic);
    CU_add_test(suite, "set_length_min_clamp", test_set_length_min_clamp);

    CU_add_test(suite, "sub_mode_value_basic", test_sub_mode_value_basic);
    CU_add_test(suite, "sub_mode_value_materializes_array", test_sub_mode_value_materializes_array);
    CU_add_test(suite, "sub_mode_length_expand", test_sub_mode_length_expand);
    CU_add_test(suite, "sub_mode_length_shrink", test_sub_mode_length_shrink);
    CU_add_test(suite, "sub_mode_length_clamp_min", test_sub_mode_length_clamp_min);
    CU_add_test(suite, "toggle_loop_mode_cycles", test_toggle_loop_mode_cycles);

    CU_add_test(suite, "chord_stack_increase", test_chord_stack_increase);
    CU_add_test(suite, "chord_stack_decrease_clamp", test_chord_stack_decrease_clamp);
    CU_add_test(suite, "chord_stack_max_clamp", test_chord_stack_max_clamp);

    CU_add_test(suite, "copy_pattern", test_copy_pattern);
    CU_add_test(suite, "clear_pattern", test_clear_pattern);
}

File: src/wasm-rust/src/test_edit.rs
// test_edit.rs — Tests for engine_edit functions
// Mirrors src/wasm/tests/test_edit.c

use crate::engine_core::*;
use crate::engine_edit::*;

fn init_state() -> Box<EngineState> {
    let mut s = EngineState::new_boxed();
    engine_core_init(&mut s);
    s
}

fn ch(s: &EngineState) -> usize { s.current_channel as usize }
fn pat(s: &EngineState) -> usize { s.current_patterns[ch(s)] as usize }

fn ev<'a>(s: &'a EngineState, idx: usize) -> &'a NoteEvent {
    let (c, p) = (ch(s), pat(s));
    let h = s.patterns[c][p].event_handles[idx];
    &s.event_pool.slots[h as usize]
}

fn ev_ch_pat<'a>(s: &'a EngineState, c: usize, p: usize, idx: usize) -> &'a NoteEvent {
    let h = s.patterns[c][p].event_handles[idx];
    &s.event_pool.slots[h as usize]
}

// ============ engine_toggle_event ============

#[test]
fn toggle_add_event() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 10, 0, 120);
    assert!(idx >= 0);
    let (c, p) = (ch(&s), pat(&s));
    assert_eq!(s.patterns[c][p].event_count, 1);
    assert_eq!(ev(&s, idx as usize).row, 10);
    assert_eq!(ev(&s, idx as usize).position, 0);
    assert_eq!(ev(&s, idx as usize).length, 120);
    assert_eq!(ev(&s, idx as usize).enabled, 1);
}

#[test]
fn toggle_remove_event() {
    let mut s = init_state();
    engine_toggle_event(&mut s, 10, 0, 120);
    let idx = engine_toggle_event(&mut s, 10, 0, 120);
    assert_eq!(idx, -1);
    let (c, p) = (ch(&s), pat(&s));
    assert_eq!(s.patterns[c][p].event_count, 0);
}

#[test]
fn toggle_multiple_events() {
    let mut s = init_state();
    engine_toggle_event(&mut s, 10, 0, 120);
    engine_toggle_event(&mut s, 12, 120, 120);
    engine_toggle_event(&mut s, 14, 240, 120);
    let (c, p) = (ch(&s), pat(&s));
    assert_eq!(s.patterns[c][p].event_count, 3);
}

#[test]
fn toggle_fill_128_events() {
    let mut s = init_state();
    // Fill 8 rows × 16 columns = 128 events, each at a unique (row, tick)
    for row in 0..8i16 {
        for col in 0..16i32 {
            let idx = engine_toggle_event(&mut s, row, col * 120, 120);
            assert!(idx >= 0, "failed to add event at row={}, col={}", row, col);
        }
    }
    let (c, p) = (ch(&s), pat(&s));
    assert_eq!(s.patterns[c][p].event_count, 128);
    assert!(s.event_pool.free_count > 0, "pool should have free slots remaining");
}

#[test]
fn toggle_event_default_values() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 5, 0, 120);
    let e = ev(&s, idx as usize);
    assert_eq!(e.repeat_amount, 1);
    assert_eq!(e.chord_amount, 1);
    assert_eq!(e.chord_space, 2);
    assert_eq!(e.arp_style, ARP_CHORD);
    assert_eq!(get_sub_mode(&s.sub_mode_pool, &e.sub_mode_handles, SubModeId::Velocity as usize).values[0], 100);
    assert_eq!(get_sub_mode(&s.sub_mode_pool, &e.sub_mode_handles, SubModeId::Velocity as usize).length, 1);
    assert_eq!(get_sub_mode(&s.sub_mode_pool, &e.sub_mode_handles, SubModeId::Hit as usize).values[0], 100);
}

// ============ engine_remove_event ============

#[test]
fn remove_event_fixes_selection() {
    let mut s = init_state();
    engine_toggle_event(&mut s, 10, 0, 120);
    engine_toggle_event(&mut s, 12, 120, 120);
    s.selected_event_idx = 1;
    engine_remove_event(&mut s, 0);
    assert_eq!(s.selected_event_idx, 0);
}

#[test]
fn remove_event_clears_selection() {
    let mut s = init_state();
    engine_toggle_event(&mut s, 10, 0, 120);
    s.selected_event_idx = 0;
    engine_remove_event(&mut s, 0);
    assert_eq!(s.selected_event_idx, -1);
}

// ============ engine_move_event ============

#[test]
fn move_event_basic() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 10, 0, 120);
    engine_move_event(&mut s, idx as u16, 20, 240);
    assert_eq!(ev(&s, idx as usize).row, 20);
    assert_eq!(ev(&s, idx as usize).position, 240);
}

#[test]
fn move_event_out_of_range_ignored() {
    let mut s = init_state();
    engine_move_event(&mut s, 99, 20, 240); // should not crash
}

// ============ engine_set_event_length ============

#[test]
fn set_length_basic() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 10, 0, 120);
    engine_set_event_length(&mut s, idx as u16, 240);
    assert_eq!(ev(&s, idx as usize).length, 240);
}

#[test]
fn set_length_min_clamp() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 10, 0, 120);
    engine_set_event_length(&mut s, idx as u16, 0);
    assert_eq!(ev(&s, idx as usize).length, 1);
}

// ============ Sub-mode operations ============

#[test]
fn sub_mode_value_basic() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 10, 0, 120);
    engine_set_sub_mode_value(&mut s, idx as u16, SubModeId::Velocity as u8, 0, 80);
    let e = ev(&s, idx as usize);
    assert_eq!(get_sub_mode(&s.sub_mode_pool, &e.sub_mode_handles, SubModeId::Velocity as usize).values[0], 80);
}

#[test]
fn sub_mode_value_materializes_array() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 10, 0, 120);
    engine_set_sub_mode_value(&mut s, idx as u16, SubModeId::Velocity as u8, 2, 50);
    let e = ev(&s, idx as usize);
    let arr = get_sub_mode(&s.sub_mode_pool, &e.sub_mode_handles, SubModeId::Velocity as usize);
    assert_eq!(arr.length, 3);
    assert_eq!(arr.values[2], 50);
    assert_eq!(arr.values[1], 100); // looped from index 0
}

#[test]
fn sub_mode_length_expand() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 10, 0, 120);
    engine_set_sub_mode_length(&mut s, idx as u16, SubModeId::Velocity as u8, 4);
    let e = ev(&s, idx as usize);
    assert_eq!(get_sub_mode(&s.sub_mode_pool, &e.sub_mode_handles, SubModeId::Velocity as usize).length, 4);
}

#[test]
fn sub_mode_length_shrink() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 10, 0, 120);
    engine_set_sub_mode_length(&mut s, idx as u16, SubModeId::Velocity as u8, 4);
    engine_set_sub_mode_length(&mut s, idx as u16, SubModeId::Velocity as u8, 2);
    let e = ev(&s, idx as usize);
    assert_eq!(get_sub_mode(&s.sub_mode_pool, &e.sub_mode_handles, SubModeId::Velocity as usize).length, 2);
}

#[test]
fn sub_mode_length_clamp_min() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 10, 0, 120);
    engine_set_sub_mode_length(&mut s, idx as u16, SubModeId::Velocity as u8, 0);
    let e = ev(&s, idx as usize);
    assert_eq!(get_sub_mode(&s.sub_mode_pool, &e.sub_mode_handles, SubModeId::Velocity as usize).length, 1);
}

#[test]
fn toggle_loop_mode_cycles() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 10, 0, 120);

    assert_eq!(get_sub_mode(&s.sub_mode_pool, &ev(&s, idx as usize).sub_mode_handles, SubModeId::Velocity as usize).loop_mode, LoopMode::Continue as u8);
    engine_cycle_sub_mode_loop_mode(&mut s, idx as u16, SubModeId::Velocity as u8, true);
    assert_eq!(get_sub_mode(&s.sub_mode_pool, &ev(&s, idx as usize).sub_mode_handles, SubModeId::Velocity as usize).loop_mode, LoopMode::Reset as u8);
    engine_cycle_sub_mode_loop_mode(&mut s, idx as u16, SubModeId::Velocity as u8, true);
    assert_eq!(get_sub_mode(&s.sub_mode_pool, &ev(&s, idx as usize).sub_mode_handles, SubModeId::Velocity as usize).loop_mode, LoopMode::Fill as u8);
    engine_cycle_sub_mode_loop_mode(&mut s, idx as u16, SubModeId::Velocity as u8, true);
    assert_eq!(get_sub_mode(&s.sub_mode_pool, &ev(&s, idx as usize).sub_mode_handles, SubModeId::Velocity as usize).loop_mode, LoopMode::Continue as u8);
}

// ============ Chord operations ============

#[test]
fn chord_stack_increase() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 10, 0, 120);

    assert_eq!(ev(&s, idx as usize).chord_amount, 1);
    engine_adjust_chord_stack(&mut s, idx as u16, 1);
    assert_eq!(ev(&s, idx as usize).chord_amount, 2);
    engine_adjust_chord_stack(&mut s, idx as u16, 1);
    assert_eq!(ev(&s, idx as usize).chord_amount, 3);
}

#[test]
fn chord_stack_decrease_clamp() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 10, 0, 120);

    engine_adjust_chord_stack(&mut s, idx as u16, -1);
    assert_eq!(ev(&s, idx as usize).chord_amount, 1);
}

#[test]
fn chord_stack_max_clamp() {
    let mut s = init_state();
    let idx = engine_toggle_event(&mut s, 10, 0, 120);

    for _ in 0..20 {
        engine_adjust_chord_stack(&mut s, idx as u16, 1);
    }
    assert!(ev(&s, idx as usize).chord_amount as usize <= MAX_CHORD_SIZE);
}

// ============ Pattern operations ============

#[test]
fn copy_pattern() {
    let mut s = init_state();
    engine_toggle_event(&mut s, 10, 0, 120);
    engine_toggle_event(&mut s, 12, 120, 120);
    engine_copy_pattern(&mut s, 1);
    let c = ch(&s);
    assert_eq!(s.patterns[c][1].event_count, 2);
    assert_eq!(ev_ch_pat(&s, c, 1, 0).row, 10);
    assert_eq!(ev_ch_pat(&s, c, 1, 1).row, 12);
}

#[test]
fn clear_pattern() {
    let mut s = init_state();
    engine_toggle_event(&mut s, 10, 0, 120);
    engine_toggle_event(&mut s, 12, 120, 120);
    let (c, p) = (ch(&s), pat(&s));
    assert_eq!(s.patterns[c][p].event_count, 2);
    engine_clear_pattern(&mut s);
    assert_eq!(s.patterns[c][p].event_count, 0);
}
